### PR TITLE
[DatePicker] Handle when GridView has not render the last child

### LINKED ISSUE #2158: [MaterialCalendarGridView] NullPointerException in onDraw in 1.3.0
**Description:** I'm receiving a similar bug to https://github.com/material-components/material-components-android/issues/1338, but using version "1.3.0"

I don't have steps to reproduce, but It occurred on different devices with different android versions. Maybe there is a trick with language, most users are using 'fr'.

**Expected behavior:** expect the app will not crash

**Source code:** 

```
private fun showDatePicker(from: DateTime, to: DateTime, onSubmit: (DateTime, DateTime) -> Unit) {
        val datePicker = MaterialDatePicker.Builder.dateRangePicker()
        datePicker.setSelection(Pair(from.millis, to.millis))
        val dialog = datePicker.build()
        dialog.addOnPositiveButtonClickListener {
            onSubmit(DateTime(it.first), DateTime(it.second))
        }
        dialog.show(parentFragmentManager, dialog.javaClass.simpleName)
    }

```
**Android API versions:**  29, 28, 27, 26

**Material Library version:** 1.3.0

**Devices:** 
Huawei P20
Xiaomi Redmi 9
Samsung Galaxy S7
Samsung Galaxy S7 edge
Samsung Galaxy A41
Samsung Galaxy Note 10
Samsung Galaxy S9
OnePlus 6

**Stacktrace:**
```
Fatal Exception: java.lang.NullPointerException: Attempt to invoke virtual method 'int android.view.View.getLeft()' on a null object reference
       at com.google.android.material.datepicker.MaterialCalendarGridView.horizontalMidPoint(MaterialCalendarGridView.java)
       at com.google.android.material.datepicker.MaterialCalendarGridView.onDraw(MaterialCalendarGridView.java:221)
       at android.view.View.draw(View.java:23187)
       at android.widget.AbsListView.draw(AbsListView.java:7190)
       at android.view.View.updateDisplayListIfDirty(View.java:22062)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at androidx.recyclerview.widget.RecyclerView.drawChild(RecyclerView.java)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.draw(View.java:23190)
       at androidx.recyclerview.widget.RecyclerView.draw(RecyclerView.java)
       at android.view.View.updateDisplayListIfDirty(View.java:22062)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.updateDisplayListIfDirty(View.java:22048)
       at android.view.View.draw(View.java:22917)
       at android.view.ViewGroup.drawChild(ViewGroup.java:5230)
       at android.view.ViewGroup.dispatchDraw(ViewGroup.java:4987)
       at android.view.View.draw(View.java:23190)
       at com.android.internal.policy.DecorView.draw(DecorView.java:1154)
       at android.view.View.updateDisplayListIfDirty(View.java:22062)
       at android.view.ThreadedRenderer.updateViewTreeDisplayList(ThreadedRenderer.java:588)
       at android.view.ThreadedRenderer.updateRootDisplayList(ThreadedRenderer.java:594)
       at android.view.ThreadedRenderer.draw(ThreadedRenderer.java:667)
       at android.view.ViewRootImpl.draw(ViewRootImpl.java:4296)
       at android.view.ViewRootImpl.performDraw(ViewRootImpl.java:4080)
       at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3348)
       at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2225)
```

I going to try the latest release 1.4.0-alpha01 of the library. But I can't reproduce this issue by myself, so the only way is to receive crash information from Firebase, it's painful. Maybe someone faced the same issue, will be thankful for any help.

### PR DESCRIPTION
**Pull request by SashaCube**
#### Don't forget:

- [x] Identify the component the PR relates to in brackets in the title.
  `[DatePicker] Handle when GridView has not render the last child`
- [x] Link to GitHub issues it solves. `closes #2158`
- [x] Sign the CLA bot. You can do this once the pull request is opened.

Handle when GridView has not render the last child

Closes  #2158

### CODE REVIEW

**Review by drchen — COMMENTED** [review]
The root cause of the issue is that when the grid view gets scrolled, the first child view won't necessary hold the item of position 0. This change fixes the crash but we may need to fix the root cause here. An official fix is on its way. 

Still thanks for you to raise the issue and push a fix!

File: lib/java/com/google/android/material/datepicker/MaterialCalendarGridView.java [hunk 1]
```diff
@@ -152,7 +153,9 @@ protected final void onDraw(@NonNull Canvas canvas) {
                     : getChildAt(firstHighlightPosition - 1).getLeft();
       } else {
         dayCompute.setTimeInMillis(startItem);
-        firstHighlightPosition = monthAdapter.dayToPosition(dayCompute.get(Calendar.DAY_OF_MONTH));
+        int day = dayCompute.get(Calendar.DAY_OF_MONTH);
+        int dayPosition = monthAdapter.dayToPosition(day);
```
Comment (raajkumars) [inline lib/java/com/google/android/material/datepicker/MaterialCalendarGridView.java:157]: This logic being repeated again down below. Consider pulling the logic into a method that will accept the time in milliseconds and return the day position.

**Review by raajkumars — APPROVED** [review]

File: lib/java/com/google/android/material/datepicker/MaterialCalendarGridView.java (hunk 1 above)
Comment (drchen) [reply lib/java/com/google/android/material/datepicker/MaterialCalendarGridView.java:157]: Hey Raaj, let's not merge the pull request. I have a more comprehensive fix pending for review.

Comment (raajkumars) [reply lib/java/com/google/android/material/datepicker/MaterialCalendarGridView.java:157]: Sounds good. thanks for letting me know.

**Comment by drchen** [comment]
Closing the pull request as the issue has been solved.

**Pull request closed without merging**